R/functions.R

```diff
@@ -0,0 +1,120 @@
+#' Make Entity Names
+#' Set first author, journal, year and Pubmed ID as the Synapse entity name.
+#'
+#' @param df Data frame from \code{"easyPubMed::article_to_df()"}.
+#' @export
+make_entity_name <- function(df){
+  first <- df[1,]
+  name <- glue::glue("{first$lastname} {first$journal} {first$year} (Pubmed ID {first$pmid})")
+  name
+}
+#' Query PubMed for publication data
+#' Returns a data frame with one author per row. Error handling for
+#' publications without an author list.
+#' @param pmid PubmedId as a string.
+#' @export
+query_pm <- function(pmid) {
+  structured_pmid_query <- easyPubMed::get_pubmed_ids(pmid)
+  text_blob <- easyPubMed::fetch_pubmed_data(structured_pmid_query)
+  df <- easyPubMed::article_to_df(text_blob)
+  if (is.null(df)){
+    df <- easyPubMed::article_to_df(text_blob, getAuthors = FALSE)
+  }
+  df
+}
+#' Mutate author naming convention
+#' `easyPubMed` returns complete names. This function preserves first initial and
+#' last name.
+#' @param df Data frame that is the output of `query_pm`.
+#' @export
+collapse_names <- function(df) {
+  firstInitials <- strsplit(df$firstname, " ")
+  firstInitials <- sapply(firstInitials, function(x){
+    paste(substring(x, 1, 1), collapse = "")
+  })
+
+  df$initials <- firstInitials
+
+  df <- df %>%
+    mutate(fullname = paste0(lastname, " ", initials))
+  df
+}
+#' Restructure Query
+#' `easyPubMed` returns one row per author per publication. This function creates a
+#' comma-separated list of authors, and one row per publication.
+#' @inheritParams collapse_names
+#' @export
+collapse_pm_query <- function(df) {
+  entry <- dplyr::group_by(df, pmid, doi, title, abstract, year, journal, entity_name)
+  entry <- summarise(entry, authors = toString(fullname))
+  if (nchar(entry$authors) > 500) {
+    entry$authors <- str_trunc(entry$authors, width = 500)
+  }
+  entry
+}
+#' Wrapper for Pubmed Query
+#' This function wraps several functions:
+#' - query Pubmed
+#' - create an entity name from first author, journal, year and Id
+#' - abbreviates the author names by first initial, last name
+#' - creates one row per PubmedId
+#' @param vec A character vector of PudmedIds.
+#' @export
+query_list_pmids <- function(vec) {
+  vec <- purrr::map(vec, ~ query_pm(.))
+  vec <- purrr::map(vec, ~ make_entity_name(.))
+  vec <- purrr::map(vec, ~ collapse_names(.))
+  vec <- purrr::map(vec, ~ collapse_pm_query(.))
```

Comment:
This isn't working for me. There's a couple things going on, I think.

1. The vector sent into this function is called `vec`, but there is also assigning of a variable called the same thing, `vec`. Even if you don't need your parameter, `vec`, after the first call to `query_pm()`, it might be good to still store the output as some other variable.
2. Each time something is assigned to `vec`, it is overwriting what it was before. By the time it gets to the `collapse_names()` section, there is only a string with an entity name; the actual query dataframe with the author names is gone.